sdk/iot/provisioning/tests/cmocka/test_az_iot_provisioning_client_parser.c

```diff
@@ -245,6 +272,94 @@ test_az_iot_provisioning_client_received_topic_payload_parse_allocation_error_st
   assert_int_equal(0, az_span_size(response.registration_information.error_tracking_id));
 }
 
+static void
+test_az_iot_provisioning_client_received_topic_and_payload_parse_invalid_json_payload_fails()
+{
+  az_iot_provisioning_client client;
+  az_span received_topic = AZ_SPAN_FROM_STR("$dps/registrations/res/200/?$rid=1");
+  az_span received_payload = AZ_SPAN_FROM_STR("123");
+
+  az_iot_provisioning_client_register_response response;
+  az_result ret = az_iot_provisioning_client_parse_received_topic_and_payload(
+      &client, received_topic, received_payload, &response);
+  assert_int_equal(AZ_ERROR_PARSER_UNEXPECTED_CHAR, ret);
+}
+
+static void
+test_az_iot_provisioning_client_received_topic_and_payload_parse_operationid_not_found_fails()
+{
+  az_iot_provisioning_client client;
+  az_span received_topic = AZ_SPAN_FROM_STR("$dps/registrations/res/200/?$rid=1");
+  az_span received_payload
+      = AZ_SPAN_FROM_STR(/*\"operationId\":\"" TEST_OPERATION_ID*/
+                         "{\"status\":\"" TEST_STATUS_FAILED "\",\"registrationState\":{"
+                         "\"registrationId\":\"" TEST_REGISTRATION_ID "\","
+                         "\"createdDateTimeUtc\":\"2020-04-10T03:11:13.0276997Z\","
+                         "\"status\":\"" TEST_STATUS_FAILED "\","
+                         "\"errorCode\":400207,"
+                         "\"errorMessage\":\"" TEST_ERROR_MESSAGE_ALLOCATION "\","
+                         "\"lastUpdatedDateTimeUtc\":\"" TEST_ERROR_TIMESTAMP "\","
+                         "\"etag\":\"IjYxMDA4ZDQ2LTAwMDAtMDEwMC0wMDAwLTVlOGZlM2QxMDAwMCI=\"}}");
+
+  az_iot_provisioning_client_register_response response;
+  az_result ret = az_iot_provisioning_client_parse_received_topic_and_payload(
+      &client, received_topic, received_payload, &response);
+  assert_int_equal(AZ_ERROR_ITEM_NOT_FOUND, ret);
+}
+
+static void
+test_az_iot_provisioning_client_received_topic_and_payload_parse_registration_state_not_found_fails()
+{
+  az_iot_provisioning_client client;
+  az_span received_topic = AZ_SPAN_FROM_STR("$dps/registrations/res/200/?$rid=1");
+  az_span received_payload
+      = AZ_SPAN_FROM_STR("{\"operationId\":\"" TEST_OPERATION_ID
+                         /*"\",\"status\":\"" TEST_STATUS_FAILED */ "\",\"registrationState\":{"
```

Comment:
~Why do we have a commented out property within the JSON?~

I was initially confused by this (especially since I saw status again below, but as part of the `registrationState` object), but I get that is the point of the test and the comment highlights what's missing.

Reply:
Yeah: it's a very interesting choice of protocol schema :). Don't get me started...